<a href="https://colab.research.google.com/github/anasmita3/SURGE/blob/main/SURGE_5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [12]:
import kagglehub

path = kagglehub.dataset_download(
    "mdsalmanhossain/bangla-social-media-abuse-dataset"
)

print(path)

100%|██████████| 4.44M/4.44M [00:01<00:00, 4.39MB/s]

Extracting files...
/root/.cache/kagglehub/datasets/mdsalmanhossain/bangla-social-media-abuse-dataset/versions/1


In [13]:
import os

for root, dirs, files in os.walk(path):
    for f in files:
        print(f)

cyberbulling bangla dataset.xlsx


In [15]:
import pandas as pd
import os

file_path = os.path.join(
    path,
    "cyberbulling bangla dataset.xlsx"
)

xls = pd.ExcelFile(file_path)

print(xls.sheet_names)

['Sheet1']


In [16]:
df = pd.read_excel(
    file_path,
    sheet_name=0
)

print(df.columns.tolist())
print()
print("Rows:", len(df))

['comment', 'Category', 'Gender', 'comment react number', 'label']

Rows: 44001


In [17]:
df.head(5).T

,0,1,2,3,4
comment,আওয়ামী লীগ স্বৈরাচারী সরকার যতো দিন ক্ষমতায় থা...,ঘরে বসে শুট করতে কেমন লেগেছে? ক্যামেরাতে কে ছি...,"অরে বাবা, এই টা কোন পাগল????",১লক্ষ ৬হাজার মামলা নিয়ে ৩৬ লক্ষ নেতা কর্মীরা ১...,যানবাহন বন্ধ করার পরও যদি এরকম সমাবেশের লোক হয...
Category,Politician,Singer,Actor,Politician,Politician
Gender,male,Male,Female,Male,Male
comment react number,1.0,2.0,2.0,1.0,3.0
label,religious,not bully,not bully,not bully,troll


In [19]:
texts = (
    df["comment"]
    .astype(str)
    .tolist()
)

print("Documents:", len(texts))

Documents: 44001


In [20]:
import re
from collections import Counter

chars = 0
words = 0
vocab = Counter()

for t in texts:

    chars += len(t)

    ws = t.split()

    words += len(ws)

    vocab.update(ws)

print("Characters :", chars)
print("Words      :", words)
print("Vocabulary :", len(vocab))
print("TTR        :", len(vocab)/words)

avg_len = (
    sum(len(w) for w in vocab)
    / len(vocab)
)

print(
    "Average Word Length:",
    avg_len
)

Characters : 4209905
Words      : 731153
Vocabulary : 96307
TTR        : 0.1317193528577466
Average Word Length: 7.603206412825651


In [21]:
chars_set = set()

for t in texts:
    chars_set.update(t)

print(
    "Character inventory:",
    len(chars_set)
)

print(
    sorted(chars_set)[:100]
)

Character inventory: 502
['\n', ' ', '!', '"', '#', '$', '%', '&', "'", '(', ')', '*', '+', ',', '-', '.', '/', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', ':', ';', '<', '=', '>', '?', '@', 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', '[', '\\', ']', '^', '_', '`', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z', '{', '|', '}', '~', '¡', '¨', '©', '\xad']


In [22]:
with open(
    "social_bn.txt",
    "w",
    encoding="utf-8"
) as f:

    for t in texts:

        t = t.strip()

        if t:

            f.write(t + "\n")

In [23]:
import os

print(
    "Size MB:",
    round(
        os.path.getsize(
            "social_bn.txt"
        )/(1024**2),
        2
    )
)

Size MB: 10.48


In [24]:
from sklearn.model_selection import train_test_split

with open(
    "social_bn.txt",
    encoding="utf-8"
) as f:

    docs = [
        line.strip()
        for line in f
        if line.strip()
    ]

train_docs, test_docs = train_test_split(
    docs,
    test_size=0.10,
    random_state=42
)

print("Train:", len(train_docs))
print("Test :", len(test_docs))

Train: 39701
Test : 4412


In [25]:
with open(
    "train_social_bn.txt",
    "w",
    encoding="utf-8"
) as f:

    for doc in train_docs:
        f.write(doc + "\n")

with open(
    "test_social_bn.txt",
    "w",
    encoding="utf-8"
) as f:

    for doc in test_docs:
        f.write(doc + "\n")

print("Saved")

Saved


In [26]:
import os

print(
    "Train MB:",
    round(
        os.path.getsize(
            "train_social_bn.txt"
        )/(1024**2),
        2
    )
)

print(
    "Test MB:",
    round(
        os.path.getsize(
            "test_social_bn.txt"
        )/(1024**2),
        2
    )
)

Train MB: 9.42
Test MB: 1.06


Transliteration

In [28]:
!pip install aksharamukha
from aksharamukha import transliterate
import time


def transliterate_file(
    infile,
    outfile
):

    start = time.time()

    count = 0

    with open(
        infile,
        encoding="utf-8"
    ) as fin, open(
        outfile,
        "w",
        encoding="utf-8"
    ) as fout:

        for line in fin:

            txt = line.strip()

            if not txt:
                continue

            iso = transliterate.process(
                "Bengali",
                "ISO",
                txt
            )

            fout.write(
                iso + "\n"
            )

            count += 1

            if count % 10000 == 0:

                elapsed = (
                    time.time()
                    - start
                )

                print(
                    f"{count:,} docs | "
                    f"{count/elapsed:.2f} docs/sec"
                )

    print(
        "Completed:",
        count
    )

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 289.9/289.9 kB 2.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 183.1/183.1 kB 10.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.4/2.4 MB 17.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.6/5.6 MB 61.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.5/1.5 MB 66.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 531.9/531.9 kB 32.0 MB/s eta 0:00:00


In [29]:
transliterate_file(
    "train_social_bn.txt",
    "train_social_iso.txt"
)

transliterate_file(
    "test_social_bn.txt",
    "test_social_iso.txt"
)

10,000 docs | 678.12 docs/sec
20,000 docs | 717.50 docs/sec
30,000 docs | 729.10 docs/sec
Completed: 39701
Completed: 4412


In [30]:
import os

for f in [

    "train_social_bn.txt",
    "train_social_iso.txt",

    "test_social_bn.txt",
    "test_social_iso.txt"

]:

    print(
        f,
        ":",
        round(
            os.path.getsize(f)
            /(1024**2),
            2
        ),
        "MB"
    )

train_social_bn.txt : 9.42 MB
train_social_iso.txt : 5.37 MB
test_social_bn.txt : 1.06 MB
test_social_iso.txt : 0.6 MB


In [31]:
from collections import Counter


def corpus_stats(fname):

    chars = 0
    words = 0

    vocab = Counter()

    with open(
        fname,
        encoding="utf-8"
    ) as f:

        for line in f:

            chars += len(line)

            ws = line.split()

            words += len(ws)

            vocab.update(ws)

    return {

        "chars": chars,

        "words": words,

        "vocab": len(vocab),

        "avg_word_len":
        sum(
            len(w)
            for w in vocab
        ) / len(vocab)

    }


bn = corpus_stats(
    "train_social_bn.txt"
)

iso = corpus_stats(
    "train_social_iso.txt"
)

print(
    "Bengali:",
    bn
)

print(
    "ISO:",
    iso
)

print()

print(
    "Expansion Factor:",
    iso["chars"]
    /
    bn["chars"]
)

print(
    "Vocabulary Ratio:",
    iso["vocab"]
    /
    bn["vocab"]
)

Bengali: {'chars': 3821471, 'words': 657308, 'vocab': 89245, 'avg_word_len': 7.549386520253235}
ISO: {'chars': 4468488, 'words': 657302, 'vocab': 86769, 'avg_word_len': 8.893429681107309}

Expansion Factor: 1.1693109799865027
Vocabulary Ratio: 0.9722561488038546


In [32]:
chars = set()

with open(
    "train_social_bn.txt",
    encoding="utf-8"
) as f:

    for line in f:

        chars.update(
            line
        )

print(
    "Character inventory:",
    len(chars)
)

print(
    sorted(chars)[:100]
)

Character inventory: 485
['\n', ' ', '!', '"', '#', '$', '%', '&', "'", '(', ')', '*', '+', ',', '-', '.', '/', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', ':', ';', '<', '=', '>', '?', '@', 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', '[', '\\', ']', '^', '_', '`', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z', '{', '|', '}', '~', '¡', '©', '\xad', '®']


In [33]:
from aksharamukha import transliterate
import random

with open(
    "test_social_bn.txt",
    encoding="utf-8"
) as f:

    docs = [
        x.strip()
        for x in f
        if x.strip()
    ]

samples = random.sample(
    docs,
    100
)

correct = 0

for s in samples:

    iso = transliterate.process(
        "Bengali",
        "ISO",
        s
    )

    back = transliterate.process(
        "ISO",
        "Bengali",
        iso
    )

    if s == back:
        correct += 1

print(
    "Exact Match:",
    correct,
    "/",
    len(samples)
)

print(
    "Accuracy:",
    correct/len(samples)
)

Exact Match: 74 / 100
Accuracy: 0.74


In [34]:
import random
from aksharamukha import transliterate

with open(
    "test_social_bn.txt",
    encoding="utf-8"
) as f:

    docs = [
        x.strip()
        for x in f
        if x.strip()
    ]

samples = random.sample(
    docs,
    20
)

for i, text in enumerate(samples, 1):

    iso = transliterate.process(
        "Bengali",
        "ISO",
        text
    )

    back = transliterate.process(
        "ISO",
        "Bengali",
        iso
    )

    print("=" * 80)
    print(f"Example {i}")
    print()
    print("Original:")
    print(text)
    print()
    print("ISO:")
    print(iso)
    print()
    print("Back:")
    print(back)
    print()
    print("Match:", text == back)
    print()

Example 1

Original:
এটাই বাঙালি। দোশ করবে আল্লাহর কাছে৷ ক্ষমা চাইবে ফেসবুকে

ISO:
ēṭāi bāṅāli. dōśa karabē āllāhara kāchē৷ kṣamā cāibē phēsabukē

Back:
এটাই বাঙালি। দোশ করবে আল্লাহর কাছে৷ ক্ষমা চাইবে ফেসবুকে

Match: True

Example 2

Original:
বাতাস নাহ দেখলেও বাতাস শরীরে লাগলে গরম বা ঠান্ডা লাগে তাই বিশ্বাস করি,মা-বাবার ডিনএ টেস্টকরলেই প্রমাণ পাওয়া যাবে বা নিজের চেহারার মিল বা মা-বাবাকে সামনাসামনি দেখেই আমরা বিশ্বাস করি,সো সাফা কবির যদি কোনকিছু নাহ দেখে বা কোনকিছুর অনুভূতি নাহ পেয়ে বিশ্বাস নাহ করতে চায় এটা একান্তই তার ব্যাপার।এখানে বাকি ছাগলরা বাটহার্টেড হচ্ছে কেন তাই বুঝিনাso,we r with safa kabir.#stand_with_safa_kabir

ISO:
bātāsa nāha dēkhalēō bātāsa śarīrē lāgalē garama bā ṭhānḍā lāgē tāi biśbāsa kari,mā-bābāra ḍinaē ṭēsṭakaralēi pramāṇa pāōẏā yābē bā nijēra cēhārāra mila bā mā-bābākē sāmanāsāmani dēkhēi āmarā biśbāsa kari,sō sāphā kabira yadi kōnakichu nāha dēkhē bā kōnakichura anubhūti nāha pēẏē biśbāsa nāha karatē cāẏa ēṭā ēkānta:i tāra byāpāra.ēkhānē bāki chāgalarā bāṭahārṭēḍa

Only Failed

In [35]:
import random
from aksharamukha import transliterate

with open(
    "test_social_bn.txt",
    encoding="utf-8"
) as f:

    docs = [
        x.strip()
        for x in f
        if x.strip()
    ]

random.shuffle(docs)

shown = 0

for text in docs:

    iso = transliterate.process(
        "Bengali",
        "ISO",
        text
    )

    back = transliterate.process(
        "ISO",
        "Bengali",
        iso
    )

    if text != back:

        print("=" * 80)
        print("Original:")
        print(text)
        print()
        print("ISO:")
        print(iso)
        print()
        print("Back:")
        print(back)
        print()

        shown += 1

        if shown == 20:
            break

Original:
এটা দিয়ে কি TikTok করা যাবে........

ISO:
ēṭā diẏē ki TikTok karā yābē........

Back:
এটা দিয়ে কি তিক্তোক্ করা যাবে॥॥॥॥

Original:
নাস্তিক সাফার উপযুক্ত শাস্তি দাবি করছি। নিজেকে মুসলিম দাবি করিস কেন নাস্তিক? তুই কুলাঙ্গার বাংলার মাটিতে তোর কোন স্হান নাই।

ISO:
nāstika sāphāra upayukta śāsti dābi karachi. nijēkē musalima dābi karisa kēna nāstika? tui kulāṅgāra bāṁlāra māṭitē tōra kōna shāna nāi.

Back:
নাস্তিক সাফার উপয়ুক্ত শাস্তি দাবি করছি। নিজেকে মুসলিম দাবি করিস কেন নাস্তিক? তুই কুলাঙ্গার বাংলার মাটিতে তোর কোন স্হান নাই।

Original:
আল্লাহ তায়ালা তাঁকে সহ এদেশের প্রতিটি মানুষকে নিরাপদ রাখুন করোনা ভাইরাস থেকে

ISO:
āllāha tāẏālā tām̐kē saha ēdēśēra pratiṭi mānuṣakē nirāpada rākhuna karōnā bhāirāsa thēkē

Back:
আল্লাহ তায়ালা তাঁকে সহ এদেশের প্রতিটি মানুষকে নিরাপদ রাখুন করোনা ভাইরাস থেকে

Original:
এর একাধিক যৌন সংগি আছে এ শয়তান মেয়ে তার বাবার সাথেও অপকর্ম করতে পারে

ISO:
ēra ēkādhika yauna saṁgi āchē ē śaẏatāna mēẏē tāra bābāra sāthēō apakarma karatē pārē

Back:
এর একাধিক যৌন

Only Correct

In [36]:
import random
from aksharamukha import transliterate

with open(
    "test_social_bn.txt",
    encoding="utf-8"
) as f:

    docs = [
        x.strip()
        for x in f
        if x.strip()
    ]

random.shuffle(docs)

shown = 0

for text in docs:

    iso = transliterate.process(
        "Bengali",
        "ISO",
        text
    )

    back = transliterate.process(
        "ISO",
        "Bengali",
        iso
    )

    if text == back:

        print("=" * 80)
        print(text)
        print()
        print(iso)
        print()

        shown += 1

        if shown == 20:
            break

বাংলাদেশে ভাই শুধু নাস্তিকের ছড়া ছড়ি

bāṁlādēśē bhāi śudhu nāstikēra chaṛā chaṛi

বউ :*

ba:u :*

তাকে নিয়ে বাজে কথা না বলাই ভালো

tākē niẏē bājē kathā nā balāi bhālō

তুমিতো ভাল না

tumitō bhāla nā

এক বুক ভালোবাসা নিও

ēka buka bhālōbāsā niō

লাল কালার

lāla kālāra

পরি তোমার দুধ দেখলে আমি আমার ধোন কে সামলাতে পারি না।

pari tōmāra dudha dēkhalē āmi āmāra dhōna kē sāmalātē pāri nā.

বাটপারীর একটা শীমা থাকা ধরকার…!

bāṭapārīra ēkaṭā śīmā thākā dharakāra…!

ফকিন্নি গো,  ভালো কিছু করো ভাইরাল এমনিতেই হবে, ফলোয়ার ভিক্ষা করতে হবেনা

phakinni gō,  bhālō kichu karō bhāirāla ēmanitēi habē, phalōẏāra bhikṣā karatē habēnā

তুমি আন্টি খুব লাফাইছচাইছ একটা কথাকইয়া আরো চেলিব্রেটি হইবাআর আরো ভাইরালহবা তবে যে এই ডা ওল্টাহইয়া গেছে।  জয় কুত্তারমত তোমার এই অবস্থাহবে।  জনগণ তোমাকেপ্রত্যাখ্যান করবে ইনশা আল্লাহ।

tumi ānṭi khuba lāphāichacāicha ēkaṭā kathāka:iẏā ārō cēlibrēṭi ha:ibāāra ārō bhāirālahabā tabē yē ēi ḍā ōlṭāha:iẏā gēchē.  jaẏa kuttāramata tōmāra ēi abasthāhabē.  janagaṇa tōmākēpratyākhyāna kara

Tokenization

In [37]:
from tokenizers import Tokenizer
from tokenizers.models import BPE
from tokenizers.trainers import BpeTrainer
from tokenizers.pre_tokenizers import Whitespace

VOCABS = [
    5000,
    10000,
    15000,
    20000,
    25000,
    30000,
    35000,
    40000,
    45000,
    50000
]

for vocab_size in VOCABS:

    print(
        f"\nTraining BPE-{vocab_size}"
    )

    tokenizer = Tokenizer(
        BPE(
            unk_token="[UNK]"
        )
    )

    tokenizer.pre_tokenizer = (
        Whitespace()
    )

    trainer = BpeTrainer(
        vocab_size=vocab_size,
        special_tokens=[
            "[UNK]"
        ]
    )

    tokenizer.train(
        ["train_social_bn.txt"],
        trainer
    )

    tokenizer.save(
        f"bpe_social_bn_{vocab_size}.json"
    )

print("Finished")


Training BPE-5000

Training BPE-10000

Training BPE-15000

Training BPE-20000

Training BPE-25000

Training BPE-30000

Training BPE-35000

Training BPE-40000

Training BPE-45000

Training BPE-50000
Finished


In [38]:
from tokenizers import Tokenizer

tok = Tokenizer.from_file(
    "bpe_social_bn_5000.json"
)

words = [
    "বাংলাদেশ",
    "কলকাতা",
    "সাহিত্য",
    "কবিতা",
    "উপন্যাস"
]

for w in words:

    print("\n", w)

    print(
        tok.encode(w).tokens
    )


 বাংলাদেশ
['বাংলাদেশ']

 কলকাতা
['কল', 'কা', 'তা']

 সাহিত্য
['সা', 'হি', 'ত্য']

 কবিতা
['কবি', 'তা']

 উপন্যাস
['উপ', 'ন্যা', 'স']


In [39]:
from tokenizers import Tokenizer
import numpy as np

def evaluate_bpe(
    tokenizer_file,
    test_file,
    vocab_size
):

    tokenizer = Tokenizer.from_file(
        tokenizer_file
    )

    total_words = 0
    total_tokens = 0
    total_chars = 0

    fragmented = 0
    oov = 0

    token_counts = []

    with open(
        test_file,
        encoding="utf-8"
    ) as f:

        for line in f:

            words = line.strip().split()

            total_words += len(words)

            sent_tokens = 0

            for word in words:

                tokens = tokenizer.encode(
                    word
                ).tokens

                total_tokens += len(tokens)

                sent_tokens += len(tokens)

                total_chars += len(word)

                if len(tokens) > 1:
                    fragmented += 1

                if "[UNK]" in tokens:
                    oov += 1

            token_counts.append(
                sent_tokens
            )

    fertility = (
        total_tokens /
        total_words
    )

    cpt = (
        total_chars /
        total_tokens
    )

    compression = cpt

    avg_tokens = np.mean(
        token_counts
    )

    wfr = (
        fragmented /
        total_words
    )

    variance = np.var(
        token_counts
    )

    return {
        "vocab": vocab_size,
        "oov": oov / total_words,
        "fertility": fertility,
        "cpt": cpt,
        "compression": compression,
        "avg_tokens": avg_tokens,
        "wfr": wfr,
        "variance": variance
    }

In [40]:
VOCABS = [
    5000,
    10000,
    15000,
    20000,
    25000,
    30000,
    35000,
    40000,
    45000,
    50000
]

bpe_social_bn_results = []

for vocab_size in VOCABS:

    res = evaluate_bpe(
        f"bpe_social_bn_{vocab_size}.json",
        "test_social_bn.txt",
        vocab_size
    )

    print(res)

    bpe_social_bn_results.append(
        res
    )

print("\nFinished.")

{'vocab': 5000, 'oov': 0.00023021193039474575, 'fertility': 1.5004265691651433, 'cpt': 3.209713084053105, 'compression': 3.209713084053105, 'avg_tokens': np.float64(25.11310063463282), 'wfr': 0.2985577899654682, 'variance': np.float64(1393.9366642754574)}
{'vocab': 10000, 'oov': 0.00023021193039474575, 'fertility': 1.3731599972916244, 'cpt': 3.5071942091300876, 'compression': 3.5071942091300876, 'avg_tokens': np.float64(22.983000906618315), 'wfr': 0.2344911639244363, 'variance': np.float64(1125.1177980661823)}
{'vocab': 15000, 'oov': 0.00023021193039474575, 'fertility': 1.3225269144830387, 'cpt': 3.641467510392988, 'compression': 3.641467510392988, 'avg_tokens': np.float64(22.135539437896647), 'wfr': 0.20762407745954364, 'variance': np.float64(1032.4000334125424)}
{'vocab': 20000, 'oov': 0.00023021193039474575, 'fertility': 1.2951858622790982, 'cpt': 3.7183379860522985, 'compression': 3.7183379860522985, 'avg_tokens': np.float64(21.67792384406165), 'wfr': 0.19217279436657864, 'variance

In [41]:
from tokenizers import Tokenizer
from tokenizers.models import BPE
from tokenizers.trainers import BpeTrainer
from tokenizers.pre_tokenizers import Whitespace

VOCABS = [
    5000,
    10000,
    15000,
    20000,
    25000,
    30000,
    35000,
    40000,
    45000,
    50000
]

for vocab_size in VOCABS:

    print(
        f"\nTraining ISO BPE-{vocab_size}"
    )

    tokenizer = Tokenizer(
        BPE(
            unk_token="[UNK]"
        )
    )

    tokenizer.pre_tokenizer = (
        Whitespace()
    )

    trainer = BpeTrainer(
        vocab_size=vocab_size,
        special_tokens=[
            "[UNK]"
        ]
    )

    tokenizer.train(
        ["train_social_iso.txt"],
        trainer
    )

    tokenizer.save(
        f"bpe_social_iso_{vocab_size}.json"
    )

print("Finished")


Training ISO BPE-5000

Training ISO BPE-10000

Training ISO BPE-15000

Training ISO BPE-20000

Training ISO BPE-25000

Training ISO BPE-30000

Training ISO BPE-35000

Training ISO BPE-40000

Training ISO BPE-45000

Training ISO BPE-50000
Finished


In [42]:
from tokenizers import Tokenizer

tok = Tokenizer.from_file(
    "bpe_social_iso_5000.json"
)

words = [
    "bāṃlā",
    "kalakātā",
    "sāhitya",
    "kabitā",
    "upanyāsa"
]

for w in words:

    print("\n", w)

    print(
        tok.encode(w).tokens
    )


 bāṃlā
['bā', '[UNK]', 'lā']

 kalakātā
['kala', 'kā', 'tā']

 sāhitya
['sā', 'hi', 'tya']

 kabitā
['kabi', 'tā']

 upanyāsa
['upa', 'nyā', 'sa']


In [44]:
from tokenizers import Tokenizer
import numpy as np

def evaluate_bpe(
    tokenizer_file,
    test_file,
    vocab_size
):

    tokenizer = Tokenizer.from_file(
        tokenizer_file
    )

    total_words = 0
    total_tokens = 0
    total_chars = 0

    fragmented = 0
    oov = 0

    token_counts = []

    with open(
        test_file,
        encoding="utf-8"
    ) as f:

        for line in f:

            words = line.strip().split()

            total_words += len(words)

            sent_tokens = 0

            for word in words:

                tokens = tokenizer.encode(
                    word
                ).tokens

                total_tokens += len(tokens)

                sent_tokens += len(tokens)

                total_chars += len(word)

                if len(tokens) > 1:
                    fragmented += 1

                if "[UNK]" in tokens:
                    oov += 1

            token_counts.append(
                sent_tokens
            )

    fertility = (
        total_tokens /
        total_words
    )

    cpt = (
        total_chars /
        total_tokens
    )

    compression = cpt

    avg_tokens = np.mean(
        token_counts
    )

    wfr = (
        fragmented /
        total_words
    )

    variance = np.var(
        token_counts
    )

    return {
        "vocab": vocab_size,
        "oov": oov / total_words,
        "fertility": fertility,
        "cpt": cpt,
        "compression": compression,
        "avg_tokens": avg_tokens,
        "wfr": wfr,
        "variance": variance
    }

In [45]:
VOCABS = [
    5000,
    10000,
    15000,
    20000,
    25000,
    30000,
    35000,
    40000,
    45000,
    50000
]

bpe_social_iso_results = []

for vocab_size in VOCABS:

    res = evaluate_bpe(
        f"bpe_social_iso_{vocab_size}.json",
        "test_social_iso.txt",
        vocab_size
    )

    print(res)

    bpe_social_iso_results.append(
        res
    )

print("\nFinished.")

{'vocab': 5000, 'oov': 0.00023021816556748778, 'fertility': 1.495443034546267, 'cpt': 3.8825207374941137, 'compression': 3.8825207374941137, 'avg_tokens': np.float64(25.02901178603808), 'wfr': 0.2956272090787211, 'variance': np.float64(1356.0929933117377)}
{'vocab': 10000, 'oov': 0.00023021816556748778, 'fertility': 1.3733461533252982, 'cpt': 4.22769494734351, 'compression': 4.22769494734351, 'avg_tokens': np.float64(22.985494106980962), 'wfr': 0.2327776498787969, 'variance': np.float64(1107.2989736225854)}
{'vocab': 15000, 'oov': 0.00023021816556748778, 'fertility': 1.3261514293839634, 'cpt': 4.378149029379027, 'compression': 4.378149029379027, 'avg_tokens': np.float64(22.195602901178603), 'wfr': 0.20706092656040517, 'variance': np.float64(1024.6219842919952)}
{'vocab': 20000, 'oov': 0.00023021816556748778, 'fertility': 1.3020191487344772, 'cpt': 4.45929585521868, 'compression': 4.45929585521868, 'avg_tokens': np.float64(21.791704442429737), 'wfr': 0.19338325907668974, 'variance': np.

In [46]:
import csv

with open(
    "social_iso_bpe_results.csv",
    "w",
    newline=""
) as f:

    writer = csv.DictWriter(
        f,
        fieldnames=bpe_social_iso_results[0].keys()
    )

    writer.writeheader()

    writer.writerows(
        bpe_social_iso_results
    )

print("Saved")

Saved


In [47]:
from tokenizers import Tokenizer
from tokenizers.models import WordPiece
from tokenizers.trainers import WordPieceTrainer
from tokenizers.pre_tokenizers import Whitespace

VOCABS = [
    5000,
    10000,
    15000,
    20000,
    25000,
    30000,
    35000,
    40000,
    45000,
    50000
]

for vocab_size in VOCABS:

    print(
        f"\nTraining Bengali WordPiece-{vocab_size}"
    )

    tokenizer = Tokenizer(
        WordPiece(
            unk_token="[UNK]"
        )
    )

    tokenizer.pre_tokenizer = (
        Whitespace()
    )

    trainer = WordPieceTrainer(
        vocab_size=vocab_size,
        special_tokens=[
            "[UNK]"
        ]
    )

    tokenizer.train(
        ["train_social_bn.txt"],
        trainer
    )

    tokenizer.save(
        f"wp_social_bn_{vocab_size}.json"
    )

print("Finished")


Training Bengali WordPiece-5000

Training Bengali WordPiece-10000

Training Bengali WordPiece-15000

Training Bengali WordPiece-20000

Training Bengali WordPiece-25000

Training Bengali WordPiece-30000

Training Bengali WordPiece-35000

Training Bengali WordPiece-40000

Training Bengali WordPiece-45000

Training Bengali WordPiece-50000
Finished


In [48]:
from tokenizers import Tokenizer

tok = Tokenizer.from_file(
    "wp_social_bn_5000.json"
)

words = [
    "বাংলাদেশ",
    "কলকাতা",
    "সাহিত্য",
    "কবিতা",
    "উপন্যাস"
]

for w in words:

    print("\n", w)

    print(
        tok.encode(w).tokens
    )


 বাংলাদেশ
['বাংলাদেশ']

 কলকাতা
['কল', '##কা', '##তা']

 সাহিত্য
['সাহ', '##িত', '##্য']

 কবিতা
['কবি', '##তা']

 উপন্যাস
['উপ', '##ন্য', '##াস']


In [49]:
from tokenizers import Tokenizer
import numpy as np

def evaluate_bpe(
    tokenizer_file,
    test_file,
    vocab_size
):

    tokenizer = Tokenizer.from_file(
        tokenizer_file
    )

    total_words = 0
    total_tokens = 0
    total_chars = 0

    fragmented = 0
    oov = 0

    token_counts = []

    with open(
        test_file,
        encoding="utf-8"
    ) as f:

        for line in f:

            words = line.strip().split()

            total_words += len(words)

            sent_tokens = 0

            for word in words:

                tokens = tokenizer.encode(
                    word
                ).tokens

                total_tokens += len(tokens)

                sent_tokens += len(tokens)

                total_chars += len(word)

                if len(tokens) > 1:
                    fragmented += 1

                if "[UNK]" in tokens:
                    oov += 1

            token_counts.append(
                sent_tokens
            )

    fertility = (
        total_tokens /
        total_words
    )

    cpt = (
        total_chars /
        total_tokens
    )

    compression = cpt

    avg_tokens = np.mean(
        token_counts
    )

    wfr = (
        fragmented /
        total_words
    )

    variance = np.var(
        token_counts
    )

    return {
        "vocab": vocab_size,
        "oov": oov / total_words,
        "fertility": fertility,
        "cpt": cpt,
        "compression": compression,
        "avg_tokens": avg_tokens,
        "wfr": wfr,
        "variance": variance
    }

In [50]:
VOCABS = [
    5000,
    10000,
    15000,
    20000,
    25000,
    30000,
    35000,
    40000,
    45000,
    50000
]

wp_social_bn_results = []

for vocab_size in VOCABS:

    res = evaluate_bpe(
        f"wp_social_bn_{vocab_size}.json",
        "test_social_bn.txt",
        vocab_size
    )

    print(res)

    wp_social_bn_results.append(
        res
    )

print("\nFinished.")

{'vocab': 5000, 'oov': 0.0002708375651702891, 'fertility': 1.5702620353443022, 'cpt': 3.066965055710787, 'compression': 3.066965055710787, 'avg_tokens': np.float64(26.28195829555757), 'wfr': 0.32359672286546143, 'variance': np.float64(1552.5773444878346)}
{'vocab': 10000, 'oov': 0.0002708375651702891, 'fertility': 1.4106168325546753, 'cpt': 3.4140658749892, 'compression': 3.4140658749892, 'avg_tokens': np.float64(23.609927470534906), 'wfr': 0.24814137720901888, 'variance': np.float64(1204.5960301851294)}
{'vocab': 15000, 'oov': 0.0002708375651702891, 'fertility': 1.347200216670052, 'cpt': 3.574775843351695, 'compression': 3.574775843351695, 'avg_tokens': np.float64(22.54850407978241), 'wfr': 0.2171304759970208, 'variance': np.float64(1069.4321441810805)}
{'vocab': 20000, 'oov': 0.0002708375651702891, 'fertility': 1.313711151736746, 'cpt': 3.665903866571832, 'compression': 3.665903866571832, 'avg_tokens': np.float64(21.98798730734361), 'wfr': 0.19955311801746903, 'variance': np.float64(

In [51]:
import csv

with open(
    "social_bn_wp_results.csv",
    "w",
    newline=""
) as f:

    writer = csv.DictWriter(
        f,
        fieldnames=wp_social_bn_results[0].keys()
    )

    writer.writeheader()

    writer.writerows(
        wp_social_bn_results
    )

print("Saved")

Saved


In [52]:
from tokenizers import Tokenizer
from tokenizers.models import WordPiece
from tokenizers.trainers import WordPieceTrainer
from tokenizers.pre_tokenizers import Whitespace

VOCABS = [
    5000,10000,15000,20000,25000,
    30000,35000,40000,45000,50000
]

for vocab_size in VOCABS:

    print(f"\nTraining ISO WP-{vocab_size}")

    tokenizer = Tokenizer(
        WordPiece(
            unk_token="[UNK]"
        )
    )

    tokenizer.pre_tokenizer = (
        Whitespace()
    )

    trainer = WordPieceTrainer(
        vocab_size=vocab_size,
        special_tokens=[
            "[UNK]"
        ]
    )

    tokenizer.train(
        ["train_social_iso.txt"],
        trainer
    )

    tokenizer.save(
        f"wp_social_iso_{vocab_size}.json"
    )

print("Finished")


Training ISO WP-5000

Training ISO WP-10000

Training ISO WP-15000

Training ISO WP-20000

Training ISO WP-25000

Training ISO WP-30000

Training ISO WP-35000

Training ISO WP-40000

Training ISO WP-45000

Training ISO WP-50000
Finished


In [53]:
from tokenizers import Tokenizer

tok = Tokenizer.from_file(
    "wp_social_iso_5000.json"
)

words = [
    "bāṃlā",
    "kalakātā",
    "sāhitya",
    "kabitā",
    "upanyāsa"
]

for w in words:

    print("\n", w)

    print(
        tok.encode(w).tokens
    )


 bāṃlā
['[UNK]']

 kalakātā
['kala', '##kā', '##tā']

 sāhitya
['sā', '##hi', '##tya']

 kabitā
['kabi', '##tā']

 upanyāsa
['upa', '##n', '##yā', '##sa']


In [54]:
VOCABS = [
    5000,10000,15000,20000,25000,
    30000,35000,40000,45000,50000
]

wp_social_iso_results = []

for vocab_size in VOCABS:

    res = evaluate_bpe(
        f"wp_social_iso_{vocab_size}.json",
        "test_social_iso.txt",
        vocab_size
    )

    print(res)

    wp_social_iso_results.append(
        res
    )

print("\nFinished.")

{'vocab': 5000, 'oov': 0.00028438714570101434, 'fertility': 1.5626667388919735, 'cpt': 3.715500207986689, 'compression': 3.715500207986689, 'avg_tokens': np.float64(26.15412511332729), 'wfr': 0.3187438213507035, 'variance': np.float64(1505.5832264104572)}
{'vocab': 10000, 'oov': 0.00028438714570101434, 'fertility': 1.409774792465095, 'cpt': 4.118451134464276, 'compression': 4.118451134464276, 'avg_tokens': np.float64(23.595194922937445), 'wfr': 0.24610321899164445, 'variance': np.float64(1183.9358608640903)}
{'vocab': 15000, 'oov': 0.00028438714570101434, 'fertility': 1.3505410126890836, 'cpt': 4.299083507139419, 'compression': 4.299083507139419, 'avg_tokens': np.float64(22.603807796917497), 'wfr': 0.2168655119645735, 'variance': np.float64(1067.1911731706737)}
{'vocab': 20000, 'oov': 0.00028438714570101434, 'fertility': 1.3191230042116382, 'cpt': 4.40147626478318, 'compression': 4.40147626478318, 'avg_tokens': np.float64(22.077969174977333), 'wfr': 0.19928767791124413, 'variance': np.

In [55]:
import csv

with open(
    "social_iso_wp_results.csv",
    "w",
    newline=""
) as f:

    writer = csv.DictWriter(
        f,
        fieldnames=wp_social_iso_results[0].keys()
    )

    writer.writeheader()

    writer.writerows(
        wp_social_iso_results
    )

print("Saved")

Saved


In [56]:
from tokenizers import Tokenizer
from tokenizers.models import Unigram
from tokenizers.trainers import UnigramTrainer
from tokenizers.pre_tokenizers import Whitespace

VOCABS = [
    5000,
    10000,
    15000,
    20000,
    25000,
    30000,
    35000,
    40000,
    45000,
    50000
]

for vocab_size in VOCABS:

    print(
        f"\nTraining Bengali Unigram-{vocab_size}"
    )

    tokenizer = Tokenizer(
        Unigram()
    )

    tokenizer.pre_tokenizer = (
        Whitespace()
    )

    trainer = UnigramTrainer(
        vocab_size=vocab_size,
        special_tokens=[
            "[UNK]"
        ]
    )

    tokenizer.train(
        ["train_social_bn.txt"],
        trainer
    )

    tokenizer.save(
        f"uni_social_bn_{vocab_size}.json"
    )

print("Finished")


Training Bengali Unigram-5000

Training Bengali Unigram-10000

Training Bengali Unigram-15000

Training Bengali Unigram-20000

Training Bengali Unigram-25000

Training Bengali Unigram-30000

Training Bengali Unigram-35000

Training Bengali Unigram-40000

Training Bengali Unigram-45000

Training Bengali Unigram-50000
Finished


In [57]:
from tokenizers import Tokenizer

tok = Tokenizer.from_file(
    "uni_social_bn_5000.json"
)

words = [
    "বাংলাদেশ",
    "কলকাতা",
    "সাহিত্য",
    "কবিতা",
    "উপন্যাস"
]

for w in words:

    print("\n", w)

    print(
        tok.encode(w).tokens
    )


 বাংলাদেশ
['বাংলাদেশ']

 কলকাতা
['কলকাতা']

 সাহিত্য
['সাহিত্য']

 কবিতা
['কবিতা']

 উপন্যাস
['উপ', 'ন্যা', 'স']


In [62]:
from tokenizers import Tokenizer
from tokenizers.models import Unigram
from tokenizers.trainers import UnigramTrainer
from tokenizers.pre_tokenizers import Whitespace

VOCABS = [
    5000,10000,15000,20000,25000,
    30000,35000,40000,45000,50000
]

for vocab_size in VOCABS:

    print(f"\nTraining BN Unigram-{vocab_size}")

    tokenizer = Tokenizer(
        Unigram()
    )

    tokenizer.pre_tokenizer = Whitespace()

    trainer = UnigramTrainer(
        vocab_size=vocab_size,
        unk_token="[UNK]",
        special_tokens=["[UNK]"]
    )

    tokenizer.train(
        ["train_social_bn.txt"],
        trainer
    )

    tokenizer.save(
        f"uni_social_bn_{vocab_size}.json"
    )

print("Finished")


Training BN Unigram-5000

Training BN Unigram-10000

Training BN Unigram-15000

Training BN Unigram-20000

Training BN Unigram-25000

Training BN Unigram-30000

Training BN Unigram-35000

Training BN Unigram-40000

Training BN Unigram-45000

Training BN Unigram-50000
Finished


In [63]:
def evaluate_unigram(
    tokenizer_file,
    test_file,
    vocab_size
):

    from tokenizers import Tokenizer
    import numpy as np

    tokenizer = Tokenizer.from_file(
        tokenizer_file
    )

    total_words = 0
    total_tokens = 0
    total_chars = 0

    fragmented = 0
    token_counts = []

    with open(
        test_file,
        encoding="utf-8"
    ) as f:

        for line in f:

            words = line.strip().split()

            total_words += len(words)

            sent_tokens = 0

            for word in words:

                try:
                    tokens = tokenizer.encode(
                        word
                    ).tokens

                except:
                    continue

                total_tokens += len(tokens)
                sent_tokens += len(tokens)
                total_chars += len(word)

                if len(tokens) > 1:
                    fragmented += 1

            token_counts.append(
                sent_tokens
            )

    fertility = total_tokens / total_words
    cpt = total_chars / total_tokens
    compression = cpt

    return {
        "vocab": vocab_size,
        "oov": 0.0,
        "fertility": fertility,
        "cpt": cpt,
        "compression": compression,
        "avg_tokens": np.mean(token_counts),
        "wfr": fragmented / total_words,
        "variance": np.var(token_counts)
    }

In [64]:
VOCABS = [
    5000,
    10000,
    15000,
    20000,
    25000,
    30000,
    35000,
    40000,
    45000,
    50000
]

uni_social_bn_results = []

for vocab_size in VOCABS:

    res = evaluate_unigram(
        f"uni_social_bn_{vocab_size}.json",
        "test_social_bn.txt",
        vocab_size
    )

    print(res)

    uni_social_bn_results.append(
        res
    )

print("\nFinished.")

{'vocab': 5000, 'oov': 0.0, 'fertility': 1.5219581555961812, 'cpt': 3.164304335833578, 'compression': 3.164304335833578, 'avg_tokens': np.float64(25.47348141432457), 'wfr': 0.3109215248154919, 'variance': np.float64(1417.4356068280772)}
{'vocab': 10000, 'oov': 0.0, 'fertility': 1.426217076308484, 'cpt': 3.3767221489000083, 'compression': 3.3767221489000083, 'avg_tokens': np.float64(23.871033544877605), 'wfr': 0.2733563545263728, 'variance': np.float64(1204.5570303914405)}
{'vocab': 15000, 'oov': 0.0, 'fertility': 1.3962353578441329, 'cpt': 3.449231366083119, 'compression': 3.449231366083119, 'avg_tokens': np.float64(23.369220308250227), 'wfr': 0.26317286207596996, 'variance': np.float64(1138.8299048317906)}
{'vocab': 20000, 'oov': 0.0, 'fertility': 1.3815559618119033, 'cpt': 3.485880357965517, 'compression': 3.485880357965517, 'avg_tokens': np.float64(23.123526745240255), 'wfr': 0.2585550815898165, 'variance': np.float64(1103.0565906445704)}
{'vocab': 25000, 'oov': 0.0, 'fertility': 1.

In [65]:
import csv

with open(
    "social_bn_unigram_results.csv",
    "w",
    newline=""
) as f:

    writer = csv.DictWriter(
        f,
        fieldnames=uni_social_bn_results[0].keys()
    )

    writer.writeheader()

    writer.writerows(
        uni_social_bn_results
    )

print("Saved")

Saved


In [69]:
from tokenizers import Tokenizer
from tokenizers.models import Unigram
from tokenizers.trainers import UnigramTrainer
from tokenizers.pre_tokenizers import Whitespace

VOCABS = [
    5000,
    10000,
    15000,
    20000,
    25000,
    30000,
    35000,
    40000,
    45000,
    50000
]

for vocab_size in VOCABS:

    print(
        f"\nTraining ISO Unigram-{vocab_size}"
    )

    tokenizer = Tokenizer(
        Unigram()
    )

    tokenizer.pre_tokenizer = (
        Whitespace()
    )

    trainer = UnigramTrainer(
        vocab_size=vocab_size,
        unk_token="[UNK]",
        special_tokens=["[UNK]"]
    )

    tokenizer.train(
        ["train_social_iso.txt"],
        trainer
    )

    tokenizer.save(
        f"uni_social_iso_{vocab_size}.json"
    )

print("Finished")


Training ISO Unigram-5000

Training ISO Unigram-10000

Training ISO Unigram-15000

Training ISO Unigram-20000

Training ISO Unigram-25000

Training ISO Unigram-30000

Training ISO Unigram-35000

Training ISO Unigram-40000

Training ISO Unigram-45000

Training ISO Unigram-50000
Finished


In [70]:
from tokenizers import Tokenizer

tok = Tokenizer.from_file(
    "uni_social_iso_5000.json"
)

words = [
    "bāṃlā",
    "kalakātā",
    "sāhitya",
    "kabitā",
    "upanyāsa"
]

for w in words:

    print("\n", w)

    print(
        tok.encode(w).tokens
    )


 bāṃlā
['bā', 'ṃ', 'lā']

 kalakātā
['kalakātā']

 sāhitya
['sā', 'hi', 'tya']

 kabitā
['kabitā']

 upanyāsa
['upa', 'n', 'yā', 'sa']


In [71]:
VOCABS = [
    5000,
    10000,
    15000,
    20000,
    25000,
    30000,
    35000,
    40000,
    45000,
    50000
]

uni_social_iso_results = []

for vocab_size in VOCABS:

    res = evaluate_unigram(
        f"uni_social_iso_{vocab_size}.json",
        "test_social_iso.txt",
        vocab_size
    )

    print(res)

    uni_social_iso_results.append(
        res
    )

print("\nFinished.")

{'vocab': 5000, 'oov': 0.0, 'fertility': 1.5524152594017036, 'cpt': 3.7400357656911067, 'compression': 3.7400357656911067, 'avg_tokens': np.float64(25.98254759746147), 'wfr': 0.3259889224435627, 'variance': np.float64(1462.114155975749)}
{'vocab': 10000, 'oov': 0.0, 'fertility': 1.4588654307111033, 'cpt': 3.9798657718120807, 'compression': 3.9798657718120807, 'avg_tokens': np.float64(24.41681776971895), 'wfr': 0.2946386251912842, 'variance': np.float64(1251.4302973983426)}
{'vocab': 15000, 'oov': 0.0, 'fertility': 1.4314830112536057, 'cpt': 4.055995459060593, 'compression': 4.055995459060593, 'avg_tokens': np.float64(23.958522212148686), 'wfr': 0.2865132781712552, 'variance': np.float64(1191.6059405178655)}
{'vocab': 20000, 'oov': 0.0, 'fertility': 1.4189564345977277, 'cpt': 4.091801870585989, 'compression': 4.091801870585989, 'avg_tokens': np.float64(23.748866727107888), 'wfr': 0.2830464634427095, 'variance': np.float64(1166.2370227410777)}
{'vocab': 25000, 'oov': 0.0, 'fertility': 1.

In [72]:
import csv

with open(
    "social_iso_unigram_results.csv",
    "w",
    newline=""
) as f:

    writer = csv.DictWriter(
        f,
        fieldnames=uni_social_iso_results[0].keys()
    )

    writer.writeheader()

    writer.writerows(
        uni_social_iso_results
    )

print("Saved")

Saved


| Model     | Script  | Fertility ↓ |      CPT ↑ |      WFR ↓ |
| --------- | ------- | ----------: | ---------: | ---------: |
| BPE       | Bengali |  **1.2471** |     3.8618 | **0.1651** |
| BPE       | ISO     |      1.2581 | **4.6150** |     0.1681 |
| WordPiece | Bengali |      1.2541 |     3.8402 |     0.1687 |
| WordPiece | ISO     |      1.2663 |     4.5851 |     0.1722 |
| Unigram   | Bengali |      1.3740 |     3.5051 |     0.2563 |
| Unigram   | ISO     |      1.4103 |     4.1168 |     0.2808 |


| Rank | Model     |  Fertility |
| ---- | --------- | ---------: |
| 1    | BPE       | **1.2471** |
| 2    | WordPiece |     1.2541 |
| 3    | Unigram   |     1.3740 |


| Rank | Model     |  Fertility |
| ---- | --------- | ---------: |
| 1    | BPE       | **1.2581** |
| 2    | WordPiece |     1.2663 |
| 3    | Unigram   |     1.4103 |


## Social Media Dataset Summary

The Social Media corpus was constructed from the Bangla Social Media Abuse Dataset containing 44,001 user-generated comments collected from online social platforms. The corpus represents informal, user-generated text and includes a wide variety of linguistic phenomena such as colloquial expressions, spelling variations, abbreviations, emotive language, code-mixing, and non-standard orthographic forms that are commonly observed in social media communication.

After preprocessing, the corpus contained 731,153 words and 4,209,905 characters with a vocabulary size of 96,307 unique word types. The Type-Token Ratio (TTR) of 0.1317 was the highest among all evaluated domains, indicating substantial lexical diversity. The average word length was 7.60 characters, reflecting the presence of both standard lexical items and elongated or expressive social-media spellings.

The corpus was divided into training and testing subsets using a 90:10 split, resulting in 39,701 training documents and 4,412 testing documents. The training corpus occupied 9.42 MB, while the testing corpus occupied 1.06 MB. To facilitate script-based comparison, the Bengali corpus was transliterated into ISO 15919 using Aksharamukha. The transliterated corpus exhibited an expansion factor of 1.1693, indicating an increase in character count due to the explicit representation of phonological distinctions in the Romanized script. The vocabulary ratio between the ISO and Bengali versions was 0.9723, suggesting minimal vocabulary loss during transliteration.

A round-trip transliteration evaluation yielded an exact-match accuracy of 74%, which was notably lower than that observed for the news and literature domains. This reduction can be attributed to the inherently noisy nature of social media text, including spelling inconsistencies, informal writing styles, punctuation variations, emojis, and other non-standard textual elements. Despite this noise, the corpus provides an important benchmark for evaluating tokenizer robustness under realistic user-generated language conditions.

Experimental results demonstrated that Byte Pair Encoding (BPE) consistently achieved the best segmentation performance across both Bengali and ISO scripts, followed closely by WordPiece, while Unigram produced comparatively higher fertility and fragmentation rates. Interestingly, despite exhibiting the highest lexical diversity among all domains, the social media corpus achieved the lowest fertility values under BPE tokenization, suggesting that the repetitive nature of user-generated comments enables efficient subword learning despite substantial vocabulary variation.
